src/queue.zig

```diff
@@ -3,126 +3,220 @@ const assert = std.debug.assert;
 
 const constants = @import("./constants.zig");
 
+const QueueLink = extern struct {
+    next: ?*QueueLink = null,
+};
+
 /// An intrusive first in/first out linked list.
 /// The element type T must have a field called "next" of type ?*T
 pub fn QueueType(comptime T: type) type {
     return struct {
-        const Queue = @This();
-
-        in: ?*T = null,
-        out: ?*T = null,
-        count: u64 = 0,
+        any: QueueAny,
 
-        // This should only be null if you're sure we'll never want to monitor `count`.
-        name: ?[]const u8,
+        pub const Link = QueueLink;
+        const Queue = @This();
 
-        // If the number of elements is large, the constants.verify check in push() can be too
-        // expensive. Allow the user to gate it. Could also be a comptime param?
-        verify_push: bool = true,
+        pub inline fn init(options: struct {
+            name: ?[]const u8,
+            verify_push: bool = true,
+        }) Queue {
+            return .{ .any = .{
+                .name = options.name,
+                .verify_push = options.verify_push,
+            } };
+        }
 
-        pub fn push(self: *Queue, elem: *T) void {
-            if (constants.verify and self.verify_push) assert(!self.contains(elem));
+        pub inline fn push(self: *Queue, link: *T) void {
+            self.any.push(&link.link);
+        }
 
-            assert(elem.next == null);
-            if (self.in) |in| {
-                in.next = elem;
-                self.in = elem;
-            } else {
-                assert(self.out == null);
-                self.in = elem;
-                self.out = elem;
-            }
-            self.count += 1;
+        pub inline fn pop(self: *Queue) ?*T {
+            const link = self.any.pop() orelse return null;
+            return @alignCast(@fieldParentPtr("link", link));
         }
 
-        pub fn pop(self: *Queue) ?*T {
-            const ret = self.out orelse return null;
-            self.out = ret.next;
-            ret.next = null;
-            if (self.in == ret) self.in = null;
-            self.count -= 1;
-            return ret;
+        pub inline fn peek_last(self: *const Queue) ?*T {
+            const link = self.any.peek_last() orelse return null;
+            return @alignCast(@fieldParentPtr("link", link));
         }
 
-        pub fn peek_last(self: Queue) ?*T {
-            return self.in;
+        pub inline fn peek(self: *const Queue) ?*T {
+            const link = self.any.peek() orelse return null;
+            return @alignCast(@fieldParentPtr("link", link));
         }
 
-        pub fn peek(self: Queue) ?*T {
-            return self.out;
+        pub fn count(self: *const Queue) u64 {
```

Comment:
I feel that we should use u32 for the size, as, if you have more than 4 billion elements in the Queue, something is very wrong. But I leave this for future work.